src/hotspot/share/opto/vtransform.cpp

```diff
@@ -781,77 +779,70 @@ VTransformApplyResult VTransformPopulateIndexNode::apply(VTransformApplyState& a
-  int   opc  = first->Opcode();
-  BasicType bt = apply_state.vloop_analyzer().types().velt_basic_type(first);
-
-  if (first->is_Cmp()) {
-    // Cmp + Bool -> VectorMaskCmp
-    // Handled by Bool / VTransformBoolVectorNode, so we do not generate any nodes here.
-    return VTransformApplyResult::make_empty();
-  }
+  uint    vlen = vector_length();
+  int     vopc = _vector_opcode;
+  BasicType bt = element_basic_type();
+  const TypeVect* vt = TypeVect::make(bt, vlen);
 
   assert(2 <= req() && req() <= 4, "Must have 1-3 inputs");
-  VectorNode* vn = nullptr;
   Node* in1 =                apply_state.transformed_node(in_req(1));
   Node* in2 = (req() >= 3) ? apply_state.transformed_node(in_req(2)) : nullptr;
   Node* in3 = (req() >= 4) ? apply_state.transformed_node(in_req(3)) : nullptr;
 
-  if (first->is_CMove()) {
-    assert(req() == 4, "three inputs expected: mask, blend1, blend2");
-    vn = new VectorBlendNode(/* blend1 */ in2, /* blend2 */ in3, /* mask */ in1);
-  } else if (VectorNode::is_convert_opcode(opc)) {
-    assert(first->req() == 2 && req() == 2, "only one input expected");
-    int vopc = VectorCastNode::opcode(opc, in1->bottom_type()->is_vect()->element_basic_type());
-    vn = VectorCastNode::make(vopc, in1, bt, vlen);
-  } else if (VectorNode::is_reinterpret_opcode(opc)) {
-    assert(first->req() == 2 && req() == 2, "only one input expected");
-    const TypeVect* vt = TypeVect::make(bt, vlen);
-    vn = new VectorReinterpretNode(in1, in1->bottom_type()->is_vect(), vt);
-  } else if (VectorNode::can_use_RShiftI_instead_of_URShiftI(first, bt)) {
-    opc = Op_RShiftI;
-    vn = VectorNode::make(opc, in1, in2, vlen, bt);
-  } else if (VectorNode::is_scalar_op_that_returns_int_but_vector_op_returns_long(opc)) {
-    // The scalar operation was a long -> int operation.
-    // However, the vector operation is long -> long.
-    VectorNode* long_vn = VectorNode::make(opc, in1, nullptr, vlen, T_LONG);
-    register_new_node_from_vectorization(apply_state, long_vn, first);
-    // Cast long -> int, to mimic the scalar long -> int operation.
-    vn = VectorCastNode::make(Op_VectorCastL2X, long_vn, T_INT, vlen);
-  } else if (req() == 3 ||
-             VectorNode::is_scalar_unary_op_with_equal_input_and_output_types(opc)) {
-    assert(!VectorNode::is_roundopD(first) || in2->is_Con(), "rounding mode must be constant");
-    vn = VectorNode::make(opc, in1, in2, vlen, bt); // unary and binary
+  VectorNode* vn = nullptr;
+  if (req() <= 3) {
+    vn = VectorNode::make(vopc, in1, in2, vt); // unary and binary
   } else {
-    assert(req() == 4, "three inputs expected");
-    assert(opc == Op_FmaD  ||
-           opc == Op_FmaF  ||
-           opc == Op_FmaHF ||
-           opc == Op_SignumF ||
-           opc == Op_SignumD,
-           "element wise operation must be from this list");
-    vn = VectorNode::make(opc, in1, in2, in3, vlen, bt); // ternary
+    vn = VectorNode::make(vopc, in1, in2, in3, vt); // ternary
```

Comment:
Moved to `SuperWordVTransformBuilder::build_inputs_for_vector_vtnodes`, to simplify the logic here.

Reply:
`is_scalar_op_that_returns_int_but_vector_op_returns_long` moved down to `VTransformElementWiseLongOpWithCastToIntVectorNode`.

`is_reinterpret_opcode` moved down to `VTransformReinterpretVectorNode`.